internal/dialectadapter/locker.go

```diff
@@ -0,0 +1,139 @@
+package dialectadapter
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/pressly/goose/v4/internal/dialectadapter/dialectquery"
+	"github.com/sethvargo/go-retry"
+)
+
+var (
+	// defaultLockID is the id used to lock the database for migrations. It is a crc64 hash of the
+	// string "goose". This is used to ensure that the lock is unique to goose.
+	//
+	// crc64.Checksum([]byte("goose"), crc64.MakeTable(crc64.ECMA))
+	defaultLockID int64 = 5887940537704921958
+
+	// ErrLockNotImplemented is returned when the database does not support locking.
+	ErrLockNotImplemented = errors.New("lock not implemented")
+)
+
+// Locker defines the methods to lock and unlock the database.
+//
+// Locking is an experimental feature and the underlying implementation may change in the future.
+//
+// The only database that currently supports locking is Postgres. Other databases will return
+// ErrLockNotImplemented.
+type Locker interface {
```

Comment:
Any plans to allow to use user-provided `Locker`? 

I can imagine that someone may want to implement a custom one (for example Redis based) locker.